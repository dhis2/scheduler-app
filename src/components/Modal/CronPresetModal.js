import React, { useState } from 'react'
import PropTypes from 'prop-types'
import {
    Button,
    Modal,
    ModalTitle,
    ModalContent,
    ModalActions,
    ButtonStrip,
    Radio,
} from '@dhis2/ui'
import i18n from '@dhis2/d2-i18n'

const cronPresets = [
    {
        label: i18n.t('Every hour'),
        value: '0 0 * ? * *',
    },
    {
        label: i18n.t('Every day at midnight'),
        value: '0 0 1 ? * *',
    },
    {
        label: i18n.t('Every day at 3 am'),
        value: '0 0 3 ? * *',
    },
    {
        label: i18n.t('Every day at noon'),
        value: '0 0 12 ? * MON-FRI',
    },
    {
        label: i18n.t('Every week'),
        value: '0 0 3 ? * MON',
    },
]

const CronPresetModal = ({ setCron, hideModal }) => {
    const [currentPreset, setCurrentPreset] = useState('')

    return (
        <Modal open small onClose={hideModal}>
            <ModalTitle>{i18n.t('Choose a preset time/interval')}</ModalTitle>
            <ModalContent>
                {cronPresets.map((preset) => (
                    <Radio
                        {...preset}
                        checked={currentPreset === preset.value}
                        key={preset.value}
                        onChange={({ value }) => setCurrentPreset(value)}
                    />
                ))}
            </ModalContent>
            <ModalActions>
                <ButtonStrip end>
                    <Button secondary onClick={hideModal} name="hide-modal">
                        {i18n.t('Cancel')}
                    </Button>
                    <Button
                        primary
                        name="insert-preset"
                        disabled={!currentPreset}
                        onClick={() => {
                            hideModal()
                            setCron(currentPreset)
                        }}
                    >
                        {i18n.t('Insert preset')}
                    </Button>
                </ButtonStrip>
            </ModalActions>
        </Modal>
    )
}

const { func } = PropTypes

CronPresetModal.propTypes = {
    hideModal: func.isRequired,
    setCron: func.isRequired,
}

export default CronPresetModal
