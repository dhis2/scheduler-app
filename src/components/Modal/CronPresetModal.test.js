import React from 'react'
import { shallow, mount } from 'enzyme'
import CronPresetModal from './CronPresetModal'

describe('<CronPresetModal>', () => {
    it('renders correctly', () => {
        const props = {
            hideModal: () => {},
            setCron: () => {},
        }
        const wrapper = shallow(<CronPresetModal {...props} />)

        expect(wrapper).toMatchSnapshot()
    })

    it('calls hideModal when cancel button is clicked', () => {
        const props = {
            hideModal: jest.fn(),
            setCron: () => {},
        }
        const wrapper = mount(<CronPresetModal {...props} />)

        wrapper.find('button[name="hide-modal"]').simulate('click')

        expect(props.hideModal).toHaveBeenCalled()
    })

    it('calls setCron and hideModal when a value is selected and insert preset button is clicked', () => {
        // Value from the presets in CronPresetModal, the test will break if this value does not exist
        const value = '0 0 3 ? * MON'
        const props = {
            hideModal: jest.fn(),
            setCron: jest.fn(),
        }
        const wrapper = mount(<CronPresetModal {...props} />)

        wrapper
            .find(`input[value="${value}"]`)
            .simulate('change', { target: { value } })
        wrapper.find('button[name="insert-preset"]').simulate('click')

        expect(props.setCron).toHaveBeenCalledWith(value)
        expect(props.hideModal).toHaveBeenCalled()
    })

    it('calls hideModal when cover is clicked', () => {
        const props = {
            hideModal: jest.fn(),
            setCron: () => {},
        }
        const wrapper = mount(<CronPresetModal {...props} />)

        wrapper.find('div[data-test="dhis2-uicore-layer"]').simulate('click')

        expect(props.hideModal).toHaveBeenCalled()
    })
})
