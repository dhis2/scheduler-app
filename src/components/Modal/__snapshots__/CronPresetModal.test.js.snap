// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CronPresetModal> renders correctly 1`] = `
<Modal
  dataTest="dhis2-uicore-modal"
  onClose={[Function]}
  open={true}
  position="top"
  small={true}
>
  <ModalTitle
    dataTest="dhis2-uicore-modaltitle"
  >
    Choose a preset time/interval
  </ModalTitle>
  <ModalContent
    dataTest="dhis2-uicore-modalcontent"
  >
    <Radio
      checked={false}
      dataTest="dhis2-uicore-radio"
      key="0 0 * ? * *"
      label="Every hour"
      onChange={[Function]}
      value="0 0 * ? * *"
    />
    <Radio
      checked={false}
      dataTest="dhis2-uicore-radio"
      key="0 0 1 ? * *"
      label="Every day at midnight"
      onChange={[Function]}
      value="0 0 1 ? * *"
    />
    <Radio
      checked={false}
      dataTest="dhis2-uicore-radio"
      key="0 0 3 ? * *"
      label="Every day at 3 am"
      onChange={[Function]}
      value="0 0 3 ? * *"
    />
    <Radio
      checked={false}
      dataTest="dhis2-uicore-radio"
      key="0 0 12 ? * MON-FRI"
      label="Every day at noon"
      onChange={[Function]}
      value="0 0 12 ? * MON-FRI"
    />
    <Radio
      checked={false}
      dataTest="dhis2-uicore-radio"
      key="0 0 3 ? * MON"
      label="Every week"
      onChange={[Function]}
      value="0 0 3 ? * MON"
    />
  </ModalContent>
  <ModalActions
    dataTest="dhis2-uicore-modalactions"
  >
    <ButtonStrip
      dataTest="dhis2-uicore-buttonstrip"
      end={true}
    >
      <Button
        dataTest="dhis2-uicore-button"
        name="hide-modal"
        onClick={[Function]}
        secondary={true}
        type="button"
      >
        Cancel
      </Button>
      <Button
        dataTest="dhis2-uicore-button"
        disabled={true}
        name="insert-preset"
        onClick={[Function]}
        primary={true}
        type="button"
      >
        Insert preset
      </Button>
    </ButtonStrip>
  </ModalActions>
</Modal>
`;
